### PR TITLE
Fix an OOB error in device-side cuvs::neighbors::refine and CAGRA kern_prune

### PR DESCRIPTION
**Pull request by achirkin**
IVF-Flat index expects all valid indices during build, which may not be the case in the context of refinement.
At the same time, `cagra::detail::graph::kern_prune` fails with OOB error if some indices are invalid.

This PR tweaks both kernels to avoid touching the input data with an invalid index.

Fixes https://github.com/rapidsai/cuvs/issues/337

### CODE REVIEW

**Comment by lijinf2** [comment]
@achirkin Thanks for the PR! It should fix the out-of-bound memory access of LONG_MAX in the `build_knn_graph` function. Since `params.compression.has_value` and `params.attach_dataset_on_build` will be executed after `build_knn_graph`, do you see any potential problem caused by the LONG_MAX in the above two functions? See https://github.com/rapidsai/cuvs/blob/7b879116684501f36ca5a19a74c01fcecb52e962/cpp/src/neighbors/detail/cagra/cagra_build.cuh#L486

**Comment by achirkin** [comment]
No, compression only affects the dataset and hence doesn't touch the graph. Also, graph optimization step should likely erase all invalid steps during truncation stage.

**Review by lowener — APPROVED** [review]
LGTM

File: cpp/test/neighbors/ann_cagra/bug_extreme_inputs_oob.cu [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+#include <cuvs/neighbors/cagra.hpp>
+
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/device_resources.hpp>
+
+#include <cstdint>
+
+namespace cuvs::neighbors::cagra {
+
+class cagra_extreme_inputs_oob_test : public ::testing::Test {
```
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/test/neighbors/ann_cagra/bug_extreme_inputs_oob.cu:30]: Are any routines from that file being used in this file? Couldn't tell just from looking at the code.

[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/test/neighbors/ann_cagra/bug_extreme_inputs_oob.cu:30]: Looks like so.. I've just tried to remove the header and compile locally and it doesn't compile.

[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/test/neighbors/ann_cagra/bug_extreme_inputs_oob.cu:29]: Got it. No biggie if it needs to remain a cu file. Was just looking at an opportunity to use cpp instead.

**Pull request merged**